### ISSUE TITLE
consult-buffer preview affects switch-to-next/previous-buffer

### ISSUE THREAD

**Issue by abcdw** [minad/consult#354]
Order of buffers for those commands is different, which is confusing. Not sure if it's a bug or it's intended.

Way to reproduce: 
1. Create buffer `test`.
2. `M-x switch-to-buffer`.
![image](https://user-images.githubusercontent.com/1218615/125035228-355ffa80-e09a-11eb-9186-61fbda73f227.png)
3. `M-x consult-buffer`.
![image](https://user-images.githubusercontent.com/1218615/125035364-63453f00-e09a-11eb-9a51-efcce78b5c2f.png)

Expected behavior: the order of first 3 buffers is the same for both commands.

**Comment by minad** [comment]
Yes, the buffers are intentionally reordered:

1. Buffers which are not currently visible in windows
2. Buffers which are visible
3. Current buffer is at the end

In case you want to change this, you can replace the function `consult--cached-all-buffers`.

https://github.com/minad/consult/blob/6d442c8e20e6a77267172801c7ac0281bd87a3a8/consult.el#L3441-L3446

**Comment by minad** [comment]
Maybe there is another issue at play here - I think `switch-to-buffer` orders by recency as determined by `buffer-name-history`. In contrast `consult-buffer` disables this recency/history-based sorting, in order to preserve the ordering of the different sources. There is no way to avoid this.

> Seems the preview functionality of consult-buffer changes the recency of buffers and this side effect makes a lot of confusion, at least for me. It is understandable from technical point of view, but doesn't seem natural from user point of view.

This should not happen. The preview passes the `norecord` option to `switch-to-buffer`.

**Comment by abcdw** [comment]
Yes, with preview disabled it doesn't happen.

Additionally checked it with the latest commit (f17db9520ddd612dc837f4112b6bcbb172acef85), the issue reproduces.

**Comment by minad** [comment]
Okay it seems `switch-to-buffer` affects the logic of `switch-to-next/previous-buffer` despite having set `norecord`. Furthermore when leaving the minibuffer, the window configuration is automatically restored (this is a feature of `read-from-minibuffer`).From what I know, everything that has to be done is done and it is still not sufficient to avoid the side effect. Maybe one could ask on the mailing list if there is a way to `(switch-to-buffer buffer 'norecord)` such that the side effect does not occur.

**Comment by minad** [comment]
@abcdw Could you give this another check with the newest commit https://github.com/minad/consult/commit/2ef56c90bc45161fa4e4b8abf573e8a846d43a58. There was the recent issue #359 which may be related to this one.

**Comment by abcdw** [comment]
Daniel Mendler ***@***.***> writes:

> @abcdw Could you give this another check with the newest commit
> https://github.com/minad/consult/commit/2ef56c90bc45161fa4e4b8abf573e8a846d43a58. There
> was the recent issue #359 which may be related to this one.

Hi Daniel!

Tried, still present :(

**Comment by minad** [comment]
Okay, it is unrelated then. Thanks for checking.

**Comment by astoff** [comment]
This is pretty easy to reproduce if you assign a key for manual buffer preview: if you start at buffer A, preview buffer B and select buffer C, then `(buffer-list)` will start with `(C A B ...)`, even of B was pretty far down the list of buffers.

Given that this reordering is pretty annoying, how about saving a copy of `(buffer-list)` before starting the previews and then call `(dolist (b (reverse saved-buffer-list)) (switch-to-buffer b))` when unwinding it?

**Comment by minad** [comment]
It would be better to figure out a way to avoid the reordering in the first place.

**Comment by astoff** [comment]
> Okay it seems switch-to-buffer affects the logic of switch-to-next/previous-buffer despite having set norecord. 

It seems that the `NORECORD` argument of `switch-to-buffer` does what it promises, but something else in the command loop records buffer changes.  Here's an experiment to confirm this:

``` elisp
(delq t
      (mapcar (lambda (other)
                (equal
                 (buffer-list)
                 (save-window-excursion
                   (switch-to-buffer other 'norecord)
                   (buffer-list))))
              (buffer-list)))
;; => nil
```

Moreover, there is basically no direct control of the buffer list exposed in Lisp.  See for instance this in `buffer.c`:

``` c
/* Alist of all buffer names vs the buffers.  This used to be
   a Lisp-visible variable, but is no longer, to prevent lossage
   due to user rplac'ing this alist or its elements.  */
Lisp_Object Vbuffer_alist;
```

So I think a manual reordering on the lines of my previous message is the best one can do.  I checked and Counsel does that as well.

**Comment by minad** [comment]
@astoff Thanks. Before going with the manual unwinding here I want to know what causes this. We should ask on the Emacs issue tracker about this. We need small reproducible examples which show the issue. Your snippet is a step in the right direction.

**Comment by oantolin** [comment]
One could also try to view the current behavior positively: you can use consult-buffer's preview to reorder your buffer list!

Another side-effect of consult's preview I've been taking advantage of is the fact that it moves the point: I've taken to using `org-store-link` as an action from `consult-outline` to quickly store links to several headings in an Org buffer. Side effects are not always evil, but a resource to be exploited. 😛

**Comment by astoff** [comment]
@oantolin So you really prefer the current behavior concerning the buffer ordering? Is a `defcustom` needed here? I have no fundamental objection to your point, but empirically I've found this reordering to be pretty maddening!

@minad I've sent a message to emacs-devel, and I'm testing this advice, which seems to do the job alright (there might be some issues when multiple frames are involved, and perhaps some subtleties concerning the `buried-buffer-list` frame parameter – I'm not quite sure what it's used for, but this code will have a side-effect on it):
``` elisp
(defun consult--restore-buffer-order (&rest args)
  (let ((buffers (buffer-list)))
    (unwind-protect
        (apply args)
      (save-window-excursion
        (dolist (buffer buffers)
          (when (buffer-live-p buffer)
            (bury-buffer-internal buffer)))))))

(advice-add 'consult--read :around 'consult--restore-buffer-order)
```

**Comment by oantolin** [comment]
> So you really prefer the current behavior concerning the buffer ordering?

No, no! I don't prefer it, @astoff. It is true that I don't *mind* it, mostly because I hardly ever rely on the buffer order, except to go back and forth between to buffers with out using `consult-buffer` in between (specially on my Chromebook which has back and forward keys ---meant for the browser--- which I bind to previous/next-buffer). In fact, I hadn't noticed that `consult-buffer` reordered the buffers. It would be better to not have the order change, I was just pointing out that it might be possible for someone to exploit the reordering.

**Comment by astoff** [comment]
[This message](https://lists.gnu.org/archive/html/emacs-devel/2022-01/msg01875.html) from emacs-devel is pretty helpful (but note that the variable `read-minibuffer-restore-windows` mentioned there is new in Emacs 28).

However... I now seem unable to reproduce the problem described [above](https://github.com/minad/consult/issues/354#issuecomment-917198655). I'll try to figure out if and when things changed when I have more time.

**Comment by minad** [comment]
@astoff I think I missed the point of this mail on emacs-devel.

> That's a different issue.  'buffer-list' is a function that returns a
list from the internal C variable Vbuffer_alist - a variable, Lisp code
is not allowed to modify directly.  In order to modify Vbuffer_alist, a
Lisp programmer has to bury buffers in precisely the same way as you do.
> You could try coming up with a macro, say 'with-buffer-list-unmodified'
(where you probably should suspend calling 'buffer-list-update-hook'
while running it).

So what does this tell me? Do we have to set `buffer-list-update-hook` to nil during preview? Do we need the manual bury-buffer-unwinding? If yes, why?

**Comment by astoff** [comment]
The conclusion from this all was:
* The `(dolist (...) (bury-buffer ...))` trick is the only way to reorder the buffer list explicitly from Lisp
* _If_ you do that, it's a good idea to unset `buffer-list-update-hook` while at it.
* But somehow the issue described in this ticket doesn't happen to me on the latest Consult release. I haven't yet got to bisecting to find out when things changed.
 
It might also be nice to hear from @abcdw, assuming he is using an up-to-date Consult.

**Comment by minad** [comment]
What is the status here? I think the issue is still present?

**Comment by minad** [comment]
@astoff Thank you for digging into this and confirming that the issue still exists. But we still don't know what causes this or if this it is reasonable that the buffer-list is modified somewhere by Emacs? If we assume that Emacs is indeed doing something reasonable here and that this is not a deeper bug, then we have no other way than adopting the bury buffer restoration technique (https://github.com/minad/consult/issues/354#issuecomment-1025236342)? I still have some hope that we can prevent this from happening in the first place or at least narrow down to what actually causes this...

**Comment by astoff** [comment]
I think I figured it out: if I do the same test, triggering some previews and then pressing `C-g`, *but I take care to go back to the empty candidate before pressing `C-g`*, then I never observe any reordering.

This is compatible with the emacs-devel answer I linked here some messages back, which mentions something being done to the buffer list when the minibuffer is unwound. So if you simulate the return to the null candidate in the exit hook (I hope it's run early enough), the issue should disappear.

**Comment by minad** [comment]
Okay, just to be very clear what you mean - we have to switch back to the original buffer, then exit the minibuffer (the unwinding), then select the actual candidate? Unfortunately this won't work since the restoration call of the Consult preview/state function is made after exiting the minibuffer. This is reasonable - if we would call restore earlier then the minibuffer unwinding would interfere. So what we would need is some pre-minibuffer-exit restore? I am not sure if we can hack something with a local minibuffer-exit-hook, maybe it is already to late then.

**Comment by astoff** [comment]
> Okay, just to be very clear what you mean - we have to switch back to the original buffer, then exit the minibuffer (the unwinding), then switch to the actually selected buffer?

Exactly.

> So what we would need is some additional pre-minibuffer-exit restoration?

I think so, yes. Note that all live preview commands should do that.

> I am not sure if we can hack something with a local minibuffer-exit-hook, maybe it is already to late then.

I hope it's not too late, this hook is still run with the minibuffer being current.

**Comment by astoff** [comment]
The following advice solves the bug to me. I'll keep using it on the lookup for adverse effects.

```elisp
(defun consult--read-hack (&rest args)
  (let* (wc
         (setup (lambda ()
                  (setq wc (current-window-configuration))))
         (unwind (lambda ()
                   (set-window-configuration wc))))
    (add-hook 'minibuffer-setup-hook setup)
    (add-hook 'minibuffer-exit-hook unwind)
    (unwind-protect
        (apply args)
      (remove-hook 'minibuffer-setup-hook setup)
      (remove-hook 'minibuffer-exit-hook unwind))))

(advice-add 'consult--read :around 'consult--read-hack)
```

**Comment by astoff** [comment]
@minad I wrote the last couple of messages on the fly as I was figuring this out, but now I'm pretty confident we should do as in https://github.com/minad/consult/issues/354#issuecomment-1069199586. I'm not going to make a pull request because I'm not sure how to best integrate with the preview mechanism, but I hope the logic is clear.

But just in case, here's what should be done, in words:
1. Before doing the first preview, remember what is the buffer of the preview window.
2. In the `minibuffer-exit-hook`, set the window buffer of the preview window to its original value before any previews took place.

**Comment by minad** [comment]
One thing is still unclear to me: Is this an Emacs bug or not? As far as I understood the process is like this:

1. We preview buffers but use norecord such that no reordering takes place (?). No issue at this point.
2. When the minibuffer exits and the currently visible buffer is not the original buffer from `minibuffer-selected-window` it will get restored with the entire window configuration.
3. The restoration of the window configuration has an effect on the buffer list. Why? (I don't think this effect makes sense.) Furthermore I wonder if this only affects the `minibuffer-selected-window`. Is this window the only one we have to restore? What happens if other windows were also changed in the meantime? Will their restoration also affect the buffer list?

**Comment by astoff** [comment]
>  1. We preview buffers but use norecord such that no reordering takes place (?). No issue at this point.

Right.

>  2. When the minibuffer exits and the currently visible buffer is not the original buffer from `minibuffer-selected-window` it will get restored with the entire window configuration.

Right.

>  3. The restoration of the window configuration has an effect on the buffer list. Why? (I don't think this effect makes sense.)

I don't know if the window configuration restoration modifies the buffer list, but this doesn't matter.

Rather, the point is that when you quit the minibuffer, the old selected window is selected again, and, for this reason, its buffer is moved to the top of the buffer list. Since we quit the minibuffer with the last previewed buffer in that window, it jumps to the top.

> Is this an Emacs bug or not?

It's an axiom that the minibuffer unwinding procedure is correct. If you changed anything in there, someone would complain it broke their workflow of 25+ years.

> Furthermore I wonder if this only affects the `minibuffer-selected-window`. Is this window the only one we have to restore? What happens if other windows were also changed in the meantime? Will their restoration also affect the buffer list?

I don't know, but, as I said, this doesn't matter, because Consult is not messing around with those other windows. Consult only restores the buffer of the minibuffer-selected-window (right?). So whatever happens  to the other windows and their buffers is correct by the axiom.

**Comment by minad** [comment]
It is not so simple. If the axiom would be a good axiom then Consult wouldn't have to do anything. Currently Consult just relies on the axiom. So clearly the axiom is insufficient. The other windows matter as soon as you leave the minibuffer to do recursive editing. Then you can also manipulate them. But after having exited the minibuffer should the buffer list be restored again? One could argue that the recursive editing is an edge case and as such doesn't matter as much as the preview when scrolling through the windows.

**Comment by astoff** [comment]
If you are doing recursive edits, then, well, you are doing something. It's only natural it will affect the buffer list.

> Currently Consult just relies on the axiom.

I think what the axiom says is that if the result of the exact sequence of steps Consult takes leads to unexpected results, then Consult is doing something wrong (since minibuffer.c isn't). Obviously the minibuffer unwinding process wasn't designed with the live-preview idea in mind. Fortunately, enough hooks are provided to make things behave as one naturally expects.

**Comment by minad** [comment]
> Obviously the minibuffer unwinding process wasn't designed with the live-preview idea in mind. Fortunately, enough hooks are provided to make things behave as one naturally expects.

This cannot be true. Why does the minibuffer even try to restore the window configuration then? The intention of the restoration procedure was to undo changes to the windows/buffers performed during minibuffer interaction. Since you can always perform recursive editing you can perform arbitrary window/buffer manipulations. The restoration was introduced with that in mind and preview is nothing special, except that it is performed automatically and except that the buffers are selected with norecord in contrast to manual buffer operations.

I think it is a perfectly fine question to ask why some things are restored and some things are not restored. Note that the situation is even worse here - we don't mess with the buffer list, we don't ask for the restoration of the buffer list. The buffer list is messed with by the minibuffer exiting procedure itself as soon as it selects the old window again! I don't see why this shouldn't count as a bug.

**Comment by astoff** [comment]
> preview is nothing special, except that it is performed automatically and except that the buffers are selected with norecord in contrast to manual buffer operations

I guess that's the difference... Or at least it explains why the questionable behavior/bug didn't surface earlier (assuming it didn't). Which I guess is your point.

We can ask, of course, but I think there's very little chance of this changing in Emacs, and even if it did, it would take 2 years before the fix is generally available.

**Comment by minad** [comment]
My point is that as soon as we discover something fishy we should try to fix the root cause. This doesn't mean that we won't add a workaround here anyway since a fix won't become available for a long time. I think that the current behavior is fishy. We exit the minibuffer with a buffer list which is unmodified in comparison to the original buffer list. However the buffer displayed in the current window has been changed. As a result we end up with a buffer list which is reordered. This doesn't sound intensional to me. But I am sure that it will be hard/impossible to convince the Emacs maintainers that there is even a bug here. And then they will be too worried to break someones space bar heating workflow. :shrug:

**Comment by astoff** [comment]
Sure, I agree. At least now we understand what's up, and we know the minimal workaround.

**Comment by minad** [comment]
So now that the preview seems to work better we should maybe revisit this issue. :)

**Comment by astoff** [comment]
Sure. I stand by the fix suggested in of https://github.com/minad/consult/issues/354#issuecomment-1069199586.

The current behavior of minibuf.c may or may not be a bug, but it can only change in Emacs 29, which will take a while to be widely available.

**Comment by oantolin** [comment]
I agree with @astoff, and don't see how a faster preview changes anything here. Just switching to the buffer is what changes the order, right? Not running the major mode hooks and stuff like that.

**Comment by minad** [comment]
The fix from https://github.com/minad/consult/issues/354#issuecomment-1069199586 is unfortunately not in a form that I want to use in Consult. But the fix itself is reasonable and minimal.

**Comment by astoff** [comment]
For the record, I'm only claiming that the advice has the correct logic to fix the problem. I'm not saying that this code is a good form to integrate the logic into Consult (that's why I didn't submit a patch). I'm sure you can do a better job than I in that regard, @minad.

**Comment by minad** [comment]
Yes, I understood you that way. I am not sure if we can find a much better approach since the window restoration has to happen right before the minibuffer is left. Consult currently doesn't have such intermediate restoration logic in place.

**Comment by astoff** [comment]
Emacs 28 introduces a variable `read-minibuffer-restore-windows`, which, when set to nil (probably a sensible setting, by the way), causes the preview buffer to no be restored when you quit `consult--read` by `C-g`. The advice fixes that problem as well.

**Comment by minad** [comment]
@astoff I know. This setting is not supported yet. We could rewrite all preview functions with the assumption of `read-minibuffer-restore-windows=nil`. This setting is also relevant for Embark (see https://github.com/oantolin/embark/issues/461).

**Comment by minad** [comment]
@astoff Maybe there is no other way than reworking the protocol of the :state function. I would then introduce a new :state2 function (better name?) for backward compatibility and phase out support of :state. The :state2 function would get two arguments ACTION and CANDIDATE. The protocol could be as follows:

- 'setup nil (inside mb)
- 'preview candidate/nil (inside mb)
- 'preview candidate/nil (inside mb)
- 'preview candidate/nil (inside mb)
- ...
- 'restore candidate/nil (inside mb)
- 'finalize candidate/nil (outside mb)

**Comment by minad** [comment]
I implemented the new API in #546. This change gives the preview/state function finer control over the setup and unwinding, such that we can fix this issue as proposed by @astoff in https://github.com/minad/consult/issues/354#issuecomment-1069199586.

**Comment by oantolin** [comment]
What's the `action` action for? Also, after calling `exit` can I call `setup` again before the minibuffer closes?

**Comment by oantolin** [comment]
I swear it used to say `action` instead of `finish`. 😄

**Comment by minad** [comment]
@oantolin 

> What's the action action for? Also, after calling exit can I call setup again before the minibuffer closes?

I will clarify the docstring of `consult--with-preview`.

**Comment by minad** [comment]
@oantolin See https://github.com/minad/consult/commit/39118aa99a392c4654c1858c1585ffb28817f0e4. Please let me know if the documentation is sufficient.

**Comment by oantolin** [comment]
The docstring is great, thanks!

**Comment by astoff** [comment]
Looks good! I'm testing the new version now.

**Comment by abcdw** [comment]
I'm on version 0.17, the behavior described in https://github.com/minad/consult/issues/354#issuecomment-877064121 still present.

**Comment by minad** [comment]
Oh no, in my tests it seemed to be gone now. Did you test emacs -Q or could this be some other side effect? If someone of you has an idea I'd like to hear it. In principle the preview restoration protocol used by Consult now should be equivalent to the one described by @astoff in https://github.com/minad/consult/issues/354#issuecomment-1069199586.

**Comment by abcdw** [comment]
emacs -q
```
M-: (require 'consult)
M-: (require 'vertico)
M-: (define-key global-map (kbd "s-b") 'consult-buffer)
M-x vertico-mode
```

```
C-x <left> [4 times]
```
Traversed buffers in this order `*scratch*`, `*Messages*`, `*GNU Emacs*`, `*scratch*`. If I press `C-x <left>` again, next will be `*Messages*`, but if I do `s-b C-g`, pressing `C-x <left>` will open `*GNU Emacs*`.

**Comment by minad** [comment]
I already checked it again. The issue seems indeed to be still present.

**Comment by minad** [comment]
I don't know. I just looked at `consult--buffer-preview` again and it seems to implement the aforementioned workaround. It previews the buffers when scrolling around and in the end it switches back to the original buffer.

https://github.com/minad/consult/blob/822928a8609730e8c22e068b04d7908312706cfd/consult.el#L4028-L4035

@astoff Do you have some insight? Maybe the problem is that I use `switch-to-buffer` here instead of `(set-window-buffer (minibuffer-selected-window) buf)` as you had it in your workaround?

**Comment by minad** [comment]
I tried to explicitly reset the window prev/next lists. But it also doesn't seem to work.

~~~ elisp
(defun consult--buffer-preview ()
  "Buffer preview function."
  ;; Only preview in current window and other window.
  ;; Preview in frames and tabs is not possible since these don't get cleaned up.
  (if (memq consult--buffer-display
            '(switch-to-buffer switch-to-buffer-other-window))
      (let ((orig-buf (current-buffer))
            (prev-bufs (window-prev-buffers))
            (next-bufs (window-next-buffers))
            other-win)
        (lambda (action cand)
          (pcase action
            ('exit
             (set-window-next-buffers nil next-bufs)
             (set-window-prev-buffers nil prev-bufs))
            ('preview
             (when (and (eq consult--buffer-display #'switch-to-buffer-other-window)
                        (not other-win))
               (switch-to-buffer-other-window orig-buf)
               (setq other-win (selected-window)))
             (let ((win (or other-win (selected-window))))
               (when (window-live-p win)
                 (with-selected-window win
                   (cond
                    ((and cand (get-buffer cand))
                     (switch-to-buffer cand 'norecord))
                    ((buffer-live-p orig-buf)
                     (switch-to-buffer orig-buf 'norecord))))))))))
    #'ignore))
~~~

I give up on this one. I simply don't know how to fix this.

**Comment by astoff** [comment]
Weird. To be honest, I don't know what `C-x <left>` is supposed to do; one would think it goes back to the previous buffers in the order shown by `C-x b`, but that's not what I observe if I follow the steps indicated above.

So it seems reasonable to drop this. At least the other issues with buffer reordering are solved.

**Comment by ctrl2wei** [comment]
I use the following advice solves this problem bug to me, hope it can provide some useful reference.

``` emacs-lisp
(defun consult-buffer-hack (&rest args)
    (if (eq consult--buffer-display #'switch-to-buffer)
        (let* ((window (selected-window))
               (orig-buffer (current-buffer))
               (orig-buffer-window-info (list orig-buffer (set-marker (make-marker) (window-start)) (point-marker)))
               (prev-buffers (window-prev-buffers window))
               (next-buffers (window-next-buffers window))
               (unwind (lambda ()
                         (set-window-prev-buffers window prev-buffers)
                         (set-window-next-buffers window next-buffers))))

          (add-hook 'minibuffer-exit-hook unwind)
          (unwind-protect
              (apply args)
            (remove-hook 'minibuffer-exit-hook unwind))

          (setq new-buffer (window-buffer window))
          (unless (eq orig-buffer new-buffer)
            (setq tmp-prev-buffer (assoc-delete-all orig-buffer prev-buffers))
            (set-window-prev-buffers window (add-to-list 'tmp-prev-buffer orig-buffer-window-info))))
      (apply args)))

  (advice-add 'consult-buffer :around 'consult-buffer-hack)

```

**Comment by minad** [comment]
@astoff Yes, I think the issue here if twofold. We have buffer-list reordering (fixed?) and we have also reordering of something else which controls the previous/next-buffer order.

@ctrl2wei This seems to be the fix I tried in https://github.com/minad/consult/issues/354#issuecomment-1120220773, but for me it didn't seem to work. Maybe I made a mistake. Anyway PR welcome if this provably fixes the issue.

**Comment by ctrl2wei** [comment]
@minad  According to the post at  [https://emacs-china.org/t/window-prev-buffers/20925/2](https://emacs-china.org/t/window-prev-buffers/20925/2), The value returned by **(window-prev-buffers)** is kind like a reference to an object。

```emacs-lisp
(set-window-prev-buffers nil prev-bufs)
```
will become like 

```emacs-lisp
 (set-window-prev-buffers nil (window-prev-buffers))
```

maybe should try another way to get the real value.

**Comment by ctrl2wei** [comment]
it seems like worked，thanks.

emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.36, cairo version 1.17.8)
emacs-repository-version： 66b4394461589d0db8690b7971000f687bd3ad57
